pump/node_test.go

```diff
@@ -168,3 +171,48 @@ func (s *heartbeatSuite) TestShouldUpdateStatus(c *C) {
 	c.Assert(p.status.MaxCommitTS, Greater, int64(0))
 	c.Assert(p.status.MaxCommitTS, LessEqual, int64(3))
 }
+
+type notifyDrainerSuite struct{}
+
+var _ = Suite(&notifyDrainerSuite{})
+
+type mockDrainerServer struct {
+	binlog.CisternServer
+	notified bool
+}
+
+func (s *mockDrainerServer) Notify(ctx context.Context, in *binlog.NotifyReq) (*binlog.NotifyResp, error) {
+	s.notified = true
+	return nil, nil
+}
+
+func (s *notifyDrainerSuite) TestNotifyDrainer(c *C) {
+	host := "127.0.0.1:8249"
+	lis, err := net.Listen("tcp", host)
+	c.Assert(err, IsNil)
+	gs := grpc.NewServer()
+	mockDrainerServerImpl := &mockDrainerServer{notified: false}
+	binlog.RegisterCisternServer(gs, mockDrainerServerImpl)
+	// defer gs.Stop()
+	go gs.Serve(lis)
+
+	cli := etcd.NewClient(testEtcdCluster.RandClient(), "drainers")
+	registry := node.NewEtcdRegistry(cli, time.Second)
+	pNode := &pumpNode{
+		EtcdRegistry: registry,
+		status:       &node.Status{State: node.Online, NodeID: "pump_notify", Addr: "http://192.168.199.100:8260"},
+		getMaxCommitTs: func() int64 {
+			return 0
+		},
+	}
+	drainerNodeStatus := &node.Status{
+		NodeID: "pump_notify_test",
+		State:  node.Online,
+		Addr:   host,
+	}
+	err = registry.UpdateNode(context.Background(), "drainers", drainerNodeStatus)
+	c.Assert(err, IsNil)
+	err = pNode.Notify(context.Background())
```

Comment:
There's a basic structure in most tests:

1. Setup dependencies
1. Run the target against some input
1. Validate the output

We can arrange our test in three paragraphs and insert blank lines between them to make it clearer.

For each paragraph, we may add comment to explain what's being setup if it's not clear from the context.

In this case, the setup step is quite a lot of work, you may want to make it simpler:

1. Is it possible to extract some functions?
1. Is the etcd setup necessary? Can we replace it with a mock?

Reply:
1. I simply the start of mock drainer as a function of `mockDrainerServer`.
2. IMO, the price of mock etcd setup is much higher than current implementation. Especially in function `pump/node.go:176:notifyDrainer` many varibles are very hard to mock.